chunked_stream/lib/src/chunked_stream_iterator.dart

```diff
@@ -23,197 +23,179 @@ import 'dart:math';
+class _ChunkedStreamIterator<T> implements ChunkedStreamIterator<T> {
+  /// Underlying iterator that iterates through the original stream.
+  final StreamIterator<List<T>> _iterator;
+
+  /// Keeps track of the number of elements left in the current substream.
+  int _toRead = 0;
+
+  /// Keeps track of the current substream we are supporting.
+  int _substreamId = 0;
+
+  /// Buffered items from a previous chunk. Items in this list should not have
+  /// been read by the user.
+  List<T> _buffered = [];
+
+  _ChunkedStreamIterator(Stream<List<T>> stream)
+      : _iterator = StreamIterator(stream);
+
+  /// Returns a list of the next [size] elements in original chunks. The size of
+  /// the chunks are not guaranteed to be consistent.
+  ///
+  /// Returns a list with a total of less than [size] elements split in chunks
+  /// if the end of stream is encounted before [size] elements are read.
+  ///
+  /// If an error is encountered before reading [size] elements, the error
+  /// will be thrown.
+  @override
+  Future<List<List<T>>> readPreservingChunks(int size) async {
+    /// Clears the remainder of elements if the user did not drain it.
+    final readToIndex = min(_toRead, _buffered.length);
+    _buffered = _buffered.sublist(readToIndex);
+    _toRead -= readToIndex;
+
+    while (_toRead > 0 && await _iterator.moveNext()) {
+      if (_toRead < _iterator.current.length) {
+        _buffered = _iterator.current.sublist(_toRead);
+        _toRead = 0;
+      } else {
+        _toRead -= _iterator.current.length;
         _buffered = [];
-        _reset();
-        return;
       }
-    };
-    _wakeup();
-    if (!c.isCompleted) {
-      _subscription.resume();
     }
 
-    return c.future;
+    final result = <List<T>>[];
+
+    /// Starts by adding from the buffer if there are buffered elements
+    /// remaining.
+    if (_buffered.isNotEmpty) {
+      final addToIndex = min(size, _buffered.length);
+      result.add(_buffered.sublist(0, addToIndex));
+      _buffered = _buffered.sublist(addToIndex);
+      size -= addToIndex;
+    }
+
+    /// Grab as many chunks as needed and add them to [result], updating
+    /// [_buffered] as necessary.
+    while (size > 0 && await _iterator.moveNext()) {
+      _buffered = _iterator.current;
+
+      final addToIndex = min(size, _buffered.length);
+      result.add(_buffered.sublist(0, addToIndex));
+      _buffered = _buffered.sublist(addToIndex);
+      size -= addToIndex;
+    }
+
+    return result;
+  }
+
+  /// Returns a list of the next [size] elements.
+  ///
+  /// Returns a list with less than [size] elements if the end of stream is
+  /// encounted before [size] elements are read.
+  ///
+  /// If an error is encountered before reading [size] elements, the error
+  /// will be thrown.
+  @override
+  Future<List<T>> read(int size) async {
+    return (await readPreservingChunks(size))
+        .expand((element) => element)
+        .toList();
   }
```

Comment:
I think this could just be:
```dart
Future<List<T>> read(int size) => readChunkedStream(substream(size));
```

using [`bufferChunkedStream`](https://pub.dev/documentation/chunked_stream/latest/chunked_stream/bufferChunkedStream.html) from `package:chunked_stream` -- unless I also made a bug in this function :see_no_evil:

Reply:
This is just a nit :D

Reply:
hehe... modifying it according to your suggestion right now! 

I'll still have to keep some of the logic for clearing the remainder of the elements if there was a previous `substream` that the user forgot to close, but your suggestion really helped made the code more concise :+1:

Reply:
Are we sure we don't want an unclosed (and uncancelled) substream to simply throw `StateError`, next time user calls `.substream()`?